containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2904 +275,1189 @@
-  dataCast1 f    = gcast1 f
+-- | /O(min(n,W))/. Find the value at a key.
+-- Calls 'error' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] ! 1    Error: element not in the map
+-- > fromList [(5,'a'), (3,'b')] ! 5 == 'a'
+(!) :: IntMap a -> Key -> a
+(!) m k = findWithDefault (error $ "IntMap.!: key " ++ show k ++ " is not an element of the map") k m
 
-fromListConstr :: Constr
-fromListConstr = mkConstr intMapDataType "fromList" [] Prefix
 
-intMapDataType :: DataType
-intMapDataType = mkDataType "Data.IntMap.Internal.IntMap" [fromListConstr]
+-- | /O(min(n,W))/. Find the value at a key.
+-- Returns 'Nothing' when the element can not be found.
+--
+-- > fromList [(5,'a'), (3,'b')] !? 1 == Nothing
+-- > fromList [(5,'a'), (3,'b')] !? 5 == Just 'a'
+--
+-- @since 0.5.11
+(!?) :: IntMap a -> Key -> Maybe a
+(!?) m k = lookup k m
 
-#endif
+-- | Same as 'difference'.
+(\\) :: IntMap a -> IntMap b -> IntMap a
+(\\) = difference
 
-{--------------------------------------------------------------------
-  Query
---------------------------------------------------------------------}
 -- | /O(1)/. Is the map empty?
 --
--- > Data.IntMap.null (empty)           == True
+-- > Data.IntMap.null empty             == True
 -- > Data.IntMap.null (singleton 1 'a') == False
-
 null :: IntMap a -> Bool
-null Nil = True
-null _   = False
-{-# INLINE null #-}
+null (IntMap Empty) = True
+null _ = False
 
 -- | /O(n)/. Number of elements in the map.
 --
 -- > size empty                                   == 0
 -- > size (singleton 1 'a')                       == 1
 -- > size (fromList([(1,'a'), (2,'c'), (3,'b')])) == 3
 size :: IntMap a -> Int
-size = go 0
-  where
-    go !acc (Bin _ _ l r) = go (go acc l) r
-    go acc (Tip _ _) = 1 + acc
-    go acc Nil = acc
+size (IntMap Empty) = 0
+size (IntMap (NonEmpty _ _ node)) = sizeNode node where
+    sizeNode :: Node t a -> Int
+    sizeNode Tip = 1
+    sizeNode (Bin _ _ l r) = sizeNode l + sizeNode r
 
 -- | /O(min(n,W))/. Is the key a member of the map?
 --
 -- > member 5 (fromList [(5,'a'), (3,'b')]) == True
 -- > member 1 (fromList [(5,'a'), (3,'b')]) == False
-
--- See Note: Local 'go' functions and capturing]
 member :: Key -> IntMap a -> Bool
-member !k = go
+member k = k `seq` start
   where
-    go (Bin p m l r) | nomatch k p m = False
-                     | zero k m  = go l
-                     | otherwise = go r
-    go (Tip kx _) = k == kx
-    go Nil = False
+    start (IntMap Empty) = False
+    start (IntMap (NonEmpty min _ node))
+        | outOfMinBound k min = False
+        | k == boundKey min = True
+        | otherwise = goL (xor k min) node
+
+    goL !_ Tip = False
+    goL !xorCache (Bin max _ l r)
+        | inMaxBound k max = if xorCache < xorCacheMax
+                    then goL xorCache l
+                    else goR xorCacheMax r
+        | outOfMaxBound k max = False
+        | otherwise = True
+      where xorCacheMax = xor k max
```

Comment:
What's meant by an xor cache? Could we collapse `goL` and `goR` into one function using the hypothetical class I described above, or would that be complicated or slow?

Reply:
Hmm, no, that class probably won't be sufficient. Hold off on trying to collapse these indefinitely.